packages/components/src/number-control/index.tsx

```diff
@@ -0,0 +1,209 @@
+	{
+		__unstableStateReducer: stateReducerProp,
+		className,
+		dragDirection = 'n',
+		hideHTMLArrows = false,
+		isDragEnabled = true,
+		isShiftStepEnabled = true,
+		label,
+		max = Infinity,
+		min = -Infinity,
+		required = false,
+		shiftStep = 10,
+		step = 1,
+		type: typeProp = 'number',
+		value: valueProp,
+		...props
+	}: WordPressComponentProps< NumberControlProps, 'input', false >,
+	ref: ForwardedRef< any >
+) {
+	const isStepAny = step === 'any';
+	const baseStep = isStepAny ? 1 : ensureNumber( step );
+	const baseValue = roundClamp( 0, min, max, baseStep );
+	const constrainValue = ( value: number, stepOverride?: number ) => {
+		// When step is "any" clamp the value, otherwise round and clamp it.
+		return isStepAny
+			? Math.min( max, Math.max( min, value ) )
+			: roundClamp( value, min, max, stepOverride ?? baseStep );
+	};
+
+	const autoComplete = typeProp === 'number' ? 'off' : undefined;
+	const classes = classNames( 'components-number-control', className );
+
+	/**
+	 * "Middleware" function that intercepts updates from InputControl.
+	 * This allows us to tap into actions to transform the (next) state for
+	 * InputControl.
+	 *
+	 * @return The updated state to apply to InputControl
+	 */
+	const numberControlStateReducer: NumberControlProps[ '__unstableStateReducer' ] =
+		( state, action ) => {
+			const nextState = { ...state };
+
+			const { type, payload } = action;
+			const event = payload.event;
+			const currentValue = nextState.value;
+
+			/**
+			 * Handles custom UP and DOWN Keyboard events
+			 */
+			if (
+				type === inputControlActionTypes.PRESS_UP ||
+				type === inputControlActionTypes.PRESS_DOWN
+			) {
+				const enableShift =
+					( event as KeyboardEvent | undefined )?.shiftKey &&
+					isShiftStepEnabled;
+
+				const incrementalValue = enableShift
+					? ensureNumber( shiftStep ) * baseStep
+					: baseStep;
+				let nextValue = isValueEmpty( currentValue )
+					? baseValue
+					: currentValue;
+
+				if ( event?.preventDefault ) {
+					event.preventDefault();
+				}
+
+				if ( type === inputControlActionTypes.PRESS_UP ) {
+					// @ts-expect-error TODO: Investigate if this is wrong
+					nextValue = add( nextValue, incrementalValue );
+				}
+
+				if ( type === inputControlActionTypes.PRESS_DOWN ) {
+					// @ts-expect-error TODO: Investigate if this is wrong
+					nextValue = subtract( nextValue, incrementalValue );
+				}
+
+				// @ts-expect-error TODO: Investigate if this is wrong
+				nextState.value = constrainValue(
+					// @ts-expect-error TODO: Investigate if this is wrong
+					nextValue,
+					enableShift ? incrementalValue : undefined
+				);
+			}
+
+			/**
+			 * Handles drag to update events
+			 */
+			if ( type === inputControlActionTypes.DRAG && isDragEnabled ) {
+				// @ts-expect-error TODO: Investigate
+				const [ x, y ] = payload.delta;
+				// @ts-expect-error TODO: Investigate
+				const enableShift = payload.shiftKey && isShiftStepEnabled;
```

Comment:
Let's leave more descriptive comments about the `expect-error` reason

Reply:
Done in 90948f0